File: android/app/src/main/java/com/marco/fittracker/data/Nutrition.kt
package com.marco.fittracker.data

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.abs

// MARK: - Energy periodization & nutrition targets (ports iOS Nutrition.swift).
// Cut/maintain/bulk: BMR (Mifflin-St Jeor) -> TDEE -> goal-adjusted target, macro
// split (ISSN), carb cycling, WHO salt cap, weight-trend regression, LEA check.

data class EnergyTargets(
    val bmr: Double, val tdee: Double, val target: Double,
    val protein: Double, val fat: Double, val carbs: Double,
    val carbHigh: Double, val carbLow: Double, val saltMax: Double,
    val rateTarget: Double, val mode: GoalMode, val adaptive: Boolean
)

data class TrendResult(
    val avgWeight: Double?, val ratePerWeek: Double?, val spanDays: Int,
    val points: Int, val status: String, val kcalAdjust: Int
)

data class LEAResult(val ea: Double?, val risk: String)

data class AdherenceResult(
    val loggingPct: Double, val avgSteps: Int?, val sessions: Int,
    val avgIntake: Int?, val days: Int, val status: String
)

fun Store.bmr(weight: Double? = null): Double {
    val w = weight ?: lastWeight
    val cm = prefs.height * 100
    val age = (prefs.age ?: 30).toDouble()
    val base = 10 * w + 6.25 * cm - 5 * age
    return if (prefs.sexCode == "f") base - 161 else base + 5
}

fun Store.tdee(weight: Double? = null): Double = bmr(weight) * prefs.activityLevel.multiplier

/** Data-driven maintenance: real maintenance ~= avg logged intake - energy
 *  implied by the measured weight trend. Null until there's enough logging. */
fun Store.adaptiveTDEE(days: Int = 21): Double? {
    val tr = weightTrend(days)
    val rate = tr.ratePerWeek ?: return null
    if (tr.points < 8) return null
    val cutoff = LocalDate.now().minusDays(days.toLong())
    val intakes = daily.mapNotNull { e ->
        val k = e.kcal ?: return@mapNotNull null
        val d = runCatching { LocalDate.parse(e.date) }.getOrNull() ?: return@mapNotNull null
        if (k > 0 && !d.isBefore(cutoff)) k else null
    }
    if (intakes.size < 8) return null
    val avgIntake = intakes.sum().toDouble() / intakes.size
    val dailyChangeKcal = rate * 7700 / 7
    return Math.round(avgIntake - dailyChangeKcal).toDouble()
}

/** Adherence over a 2-3 week window: logging consistency, steps, volume. */
fun Store.adherence(days: Int = 14): AdherenceResult {
    val cutoff = LocalDate.now().minusDays(days.toLong())
    fun recent(ds: String): Boolean {
        val d = runCatching { LocalDate.parse(ds) }.getOrNull() ?: return false
        return !d.isBefore(cutoff)
    }
    val logged = daily.filter { recent(it.date) && (it.kcal ?: 0) > 0 }
    val stepVals = daily.mapNotNull { if (recent(it.date)) it.steps?.takeIf { s -> s > 0 } else null }
    val sess = sessions.count { recent(it.date) }
    val loggingPct = logged.size.toDouble() / days
    val avgSteps = if (stepVals.isEmpty()) null else stepVals.sum() / stepVals.size
    val avgIntake = if (logged.isEmpty()) null else logged.mapNotNull { it.kcal }.sum() / logged.size
    val status = when {
        logged.isEmpty() -> "none"
        loggingPct < 0.5 -> "low_logging"
        else -> "ok"
    }
    return AdherenceResult(loggingPct, avgSteps, sess, avgIntake, days, status)
}

fun Store.targetRate(): Double {
    prefs.weeklyRate?.let { if (abs(it) > 0.001) return it }
    return prefs.goal.defaultWeeklyPct / 100 * lastWeight
}

fun Store.fatFreeMass(): Double? {
    val bf = currentBF ?: return null
    if (bf <= 0 || bf >= 60) return null
    return lastWeight * (1 - bf / 100)
}

fun Store.energyTargets(): EnergyTargets {
    val w = lastWeight
    val adaptive = adaptiveTDEE()
    val td = adaptive ?: tdee(w)
    val rate = targetRate()
    val dailyDelta = rate * 7700 / 7
    var target = td + dailyDelta
    target = maxOf(target, bmr(w) * 1.1)
    val mode = prefs.goal
    val proteinPerKg = if (mode == GoalMode.CUT) 2.2 else if (mode == GoalMode.BULK) 1.8 else 2.0
    val protein = Math.round(proteinPerKg * w).toDouble()
    val fat = Math.round(0.8 * w).toDouble()
    val carbsKcal = maxOf(0.0, target - protein * 4 - fat * 9)
    val carbs = Math.round(carbsKcal / 4).toDouble()
    return EnergyTargets(
        bmr = Math.round(bmr(w)).toDouble(),
        tdee = Math.round(td).toDouble(),
        target = Math.round(target).toDouble(),
        protein = protein, fat = fat, carbs = carbs,
        carbHigh = Math.round(carbs * 1.30).toDouble(),
        carbLow = Math.round(carbs * 0.65).toDouble(),
        saltMax = 5.0,
        rateTarget = Math.round(rate * 100.0) / 100.0,
        mode = mode,
        adaptive = adaptive != null
    )
}

fun Store.weightTrend(days: Int = 21): TrendResult {
    val cutoff = LocalDate.now().minusDays(days.toLong())
    val pts = sortedDaily.mapNotNull { e ->
        val w = e.weight ?: return@mapNotNull null
        val d = runCatching { LocalDate.parse(e.date) }.getOrNull() ?: return@mapNotNull null
        if (d.isBefore(cutoff)) null else ChronoUnit.DAYS.between(cutoff, d).toDouble() to w
    }
    if (pts.size < 4)
        return TrendResult(lastWeight, null, days, pts.size, "none", 0)
    val n = pts.size.toDouble()
    val sx = pts.sumOf { it.first }
    val sy = pts.sumOf { it.second }
    val sxx = pts.sumOf { it.first * it.first }
    val sxy = pts.sumOf { it.first * it.second }
    val denom = n * sxx - sx * sx
    if (denom == 0.0) return TrendResult(lastWeight, null, days, pts.size, "none", 0)
    val slope = (n * sxy - sx * sy) / denom
    val intercept = (sy - slope * sx) / n
    val lastX = pts.maxOf { it.first }
    val avg = intercept + slope * lastX
    val rate = slope * 7
    val target = targetRate()
    var status = "ok"
    var adjust = 0
    val band = maxOf(0.1, abs(target) * 0.4)
    if (target < 0) {
        if (rate > 0.05) { status = "wrong"; adjust = -250 }
        else if (rate > target + band) { status = "slow"; adjust = -150 }
        else if (rate < target - band) { status = "fast"; adjust = 150 }
        else status = "ok"
    } else if (target > 0) {
        if (rate < -0.05) { status = "wrong"; adjust = 250 }
        else if (rate < target - band) { status = "slow"; adjust = 150 }
        else if (rate > target + band) { status = "fast"; adjust = -150 }
        else status = "ok"
    } else {
        if (abs(rate) > band) { status = "fast"; adjust = if (rate > 0) -150 else 150 }
        else status = "ok"
    }
    return TrendResult(
        Math.round(avg * 10.0) / 10.0,
        Math.round(rate * 100.0) / 100.0,
        days, pts.size, status, adjust
    )
}

fun Store.energyAvailability(): LEAResult {
    val ffm = fatFreeMass() ?: return LEAResult(null, "none")
    if (ffm <= 0) return LEAResult(null, "none")
    val cutoff = LocalDate.now().minusDays(7)
    val intakes = daily.mapNotNull { e ->
        val k = e.kcal ?: return@mapNotNull null
        val d = runCatching { LocalDate.parse(e.date) }.getOrNull() ?: return@mapNotNull null
        if (k > 0 && !d.isBefore(cutoff)) k else null
    }
    if (intakes.isEmpty()) return LEAResult(null, "none")
    val avgIntake = intakes.sum().toDouble() / intakes.size
    val exDays = sessions.filter {
        val d = runCatching { LocalDate.parse(it.date) }.getOrNull()
        d != null && !d.isBefore(cutoff)
    }
    val exEnergy = exDays.sumOf { (it.caloriesManual ?: estimateCalories(it)).toDouble() }
    val ea = (avgIntake - exEnergy / 7) / ffm
    val risk = if (ea < 30) "risk" else if (ea < 45) "warn" else "ok"
    return LEAResult(Math.round(ea * 10.0) / 10.0, risk)
}

File: ios/FitTracker/Nutrition.swift
import Foundation

// MARK: - Energy periodization & nutrition targets
// Cut / maintain / bulk engine: BMR (Mifflin-St Jeor) -> TDEE -> goal-adjusted
// calorie target, macro split (ISSN ranges), carb cycling across high/low days,
// WHO salt cap, a body-weight trend engine, and a low-energy-availability check.

struct EnergyTargets {
    var bmr: Double
    var tdee: Double
    var target: Double          // daily kcal target
    var protein: Double         // g/day
    var fat: Double             // g/day
    var carbs: Double           // g/day (baseline)
    var carbHigh: Double        // g on training days
    var carbLow: Double         // g on rest days
    var saltMax: Double         // g/day (WHO)
    var rateTarget: Double      // kg/week (signed)
    var mode: GoalMode
}

struct TrendResult {
    var avgWeight: Double?       // smoothed current weight
    var ratePerWeek: Double?     // kg/week from regression
    var spanDays: Int
    var points: Int
    /// "ok" | "fast" | "slow" | "wrong" | "none"
    var status: String
    /// suggested daily kcal adjustment to get back on target
    var kcalAdjust: Int
}

struct LEAResult {
    var ea: Double?              // kcal per kg FFM per day
    /// "ok" | "warn" | "risk" | "none"
    var risk: String
}

extension Store {
    // --- Basal & total energy ------------------------------------------------
    func bmr(weight: Double? = nil) -> Double {
        let w = weight ?? lastWeight
        let cm = prefs.height * 100
        let age = Double(prefs.age ?? 30)
        let base = 10 * w + 6.25 * cm - 5 * age
        return prefs.sex_ == "f" ? base - 161 : base + 5
    }

    func tdee(weight: Double? = nil) -> Double {
        bmr(weight: weight) * prefs.activityLevel.multiplier
    }

    /// Target kg/week (signed). User value wins, else derived from goal mode.
    func targetRate() -> Double {
        if let r = prefs.weeklyRate, abs(r) > 0.001 { return r }
        return prefs.goal.defaultWeeklyPct / 100 * lastWeight
    }

    /// Fat-free mass from current body-fat %, if known.
    func fatFreeMass() -> Double? {
        guard let bf = currentBF, bf > 0, bf < 60 else { return nil }
        return lastWeight * (1 - bf / 100)
    }

    func energyTargets() -> EnergyTargets {
        let w = lastWeight
        let td = tdee(weight: w)
        let rate = targetRate()                         // kg/week
        let dailyDelta = rate * 7700 / 7                // ~7700 kcal per kg
        var target = td + dailyDelta
        target = max(target, bmr(weight: w) * 1.1)      // never below ~BMR
        let mode = prefs.goal

        // Macros: protein scaled by goal (higher in a cut to protect lean mass),
        // fat ~0.8 g/kg floor, carbs fill the remainder.
        let proteinPerKg: Double = mode == .cut ? 2.2 : (mode == .bulk ? 1.8 : 2.0)
        let protein = (proteinPerKg * w).rounded()
        let fat = max(0.6 * w, 0.8 * w).rounded()
        let carbsKcal = max(0, target - protein * 4 - fat * 9)
        let carbs = (carbsKcal / 4).rounded()

        // Carb cycling: shuttle carbs toward training days, keep the weekly mean.
        let carbHigh = (carbs * 1.30).rounded()
        let carbLow  = (carbs * 0.65).rounded()

        return EnergyTargets(
            bmr: bmr(weight: w).rounded(),
            tdee: td.rounded(),
            target: target.rounded(),
            protein: protein, fat: fat, carbs: carbs,
            carbHigh: carbHigh, carbLow: carbLow,
            saltMax: 5,                                 // WHO recommendation
            rateTarget: (rate * 100).rounded() / 100,
            mode: mode
        )
    }

    // --- Body-weight trend engine -------------------------------------------
    /// Linear regression over the last `days` of weigh-ins to estimate the real
    /// rate of change (kg/week), independent of daily noise.
    func weightTrend(days: Int = 21) -> TrendResult {
        let cal = Calendar.current
        let cutoff = cal.date(byAdding: .day, value: -days, to: Date())!
        let pts = sortedDaily.compactMap { e -> (Double, Double)? in
            guard let w = e.weight, let d = isoFormatter.date(from: e.date), d >= cutoff else { return nil }
            let x = d.timeIntervalSince(cutoff) / 86400         // day index
            return (x, w)
        }
        guard pts.count >= 4 else {
            return TrendResult(avgWeight: lastWeight, ratePerWeek: nil, spanDays: days, points: pts.count, status: "none", kcalAdjust: 0)
        }
        // least-squares slope (kg/day)
        let n = Double(pts.count)
        let sx = pts.reduce(0) { $0 + $1.0 }
        let sy = pts.reduce(0) { $0 + $1.1 }
        let sxx = pts.reduce(0) { $0 + $1.0 * $1.0 }
        let sxy = pts.reduce(0) { $0 + $1.0 * $1.1 }
        let denom = n * sxx - sx * sx
        guard denom != 0 else {
            return TrendResult(avgWeight: lastWeight, ratePerWeek: nil, spanDays: days, points: pts.count, status: "none", kcalAdjust: 0)
        }
        let slope = (n * sxy - sx * sy) / denom              // kg/day
        let intercept = (sy - slope * sx) / n
        let lastX = pts.map { $0.0 }.max() ?? 0
        let avg = intercept + slope * lastX                  // fitted current weight
        let rate = slope * 7                                 // kg/week

        let target = targetRate()
        let status: String
        var adjust = 0
        let band = max(0.1, abs(target) * 0.4)               // tolerance kg/week
        if target < 0 {                                      // cutting
            if rate > 0.05 { status = "wrong"; adjust = -250 }
            else if rate > target + band { status = "slow"; adjust = -150 }
            else if rate < target - band { status = "fast"; adjust = 150 }
            else { status = "ok" }
        } else if target > 0 {                               // bulking
            if rate < -0.05 { status = "wrong"; adjust = 250 }
            else if rate < target - band { status = "slow"; adjust = 150 }
            else if rate > target + band { status = "fast"; adjust = -150 }
            else { status = "ok" }
        } else {                                             // maintenance
            if abs(rate) > band { status = abs(rate) > 0 ? "fast" : "ok"; adjust = rate > 0 ? -150 : 150 }
            else { status = "ok" }
        }
        return TrendResult(avgWeight: (avg * 10).rounded() / 10,
                           ratePerWeek: (rate * 100).rounded() / 100,
                           spanDays: days, points: pts.count, status: status, kcalAdjust: adjust)
    }

    // --- Low energy availability --------------------------------------------
    /// EA = (intake - exercise energy) / FFM, averaged over the last 7 days.
    /// <30 kcal/kg FFM/day = clinical LEA risk; 30-45 a caution under heavy load.
    func energyAvailability() -> LEAResult {
        guard let ffm = fatFreeMass(), ffm > 0 else { return LEAResult(ea: nil, risk: "none") }
        let cal = Calendar.current
        let cutoff = cal.date(byAdding: .day, value: -7, to: Date())!
        let intakes = daily.compactMap { e -> Int? in
            guard let k = e.kcal, k > 0, let d = isoFormatter.date(from: e.date), d >= cutoff else { return nil }
            return k
        }
        guard !intakes.isEmpty else { return LEAResult(ea: nil, risk: "none") }
        let avgIntake = Double(intakes.reduce(0, +)) / Double(intakes.count)
        let exDays = sessions.filter { (isoFormatter.date(from: $0.date) ?? .distantPast) >= cutoff }
        let exEnergy = exDays.reduce(0.0) { $0 + Double($1.caloriesManual ?? estimateCalories($1)) }
        let ea = (avgIntake - exEnergy / 7) / ffm
        let risk: String
        if ea < 30 { risk = "risk" }
        else if ea < 45 { risk = "warn" }
        else { risk = "ok" }
        return LEAResult(ea: (ea * 10).rounded() / 10, risk: risk)
    }
}
